lib/resque/durable/background_heartbeat.rb

```diff
@@ -0,0 +1,83 @@
+require 'thread'
+
+module Resque
+  module Durable
+
+    # Creates a background thread to regularly heartbeat the queue audit.
+    class BackgroundHeartbeat
+      DEFAULT_INTERVAL = 15
+
+      def initialize(queue_audit, interval = nil)
+        @queue_audit  = queue_audit
+        @last_timeout = nil
+        @interval     = interval || DEFAULT_INTERVAL
+        @mutex        = Mutex.new
+        @cv           = ConditionVariable.new
+        @stop         = false
+        @thread       = nil
+      end
+
+      class << self
+        # only a separate method for easy stubbing
+        def exit_now!
+          abort
+        end
+      end
+
+      def with_heartbeat
+        start!
+        yield
+      ensure
+        stop_and_wait!
+      end
+
+      def heartbeat!
+        @last_timeout ||= @queue_audit.timeout_at
+        @last_timeout = @queue_audit.optimistic_heartbeat!(@last_timeout)
```

Comment:
This does not directly update the `@queue_audit` instance itself -- the new value is only in MySQL and the `BackgroundHeartbeat` instance. I could update the value in the instance, but it would mark it as changed, and I want to avoid mucking with the ActiveModel::Dirty internals. This is not he first time I've wanted ActiveRecord to expose better `UPDATE ... WHERE` primitives. The `ActiveRecord::Locking::Optimistic` implementation is unfortunately coupled tightly to internals and changes often.